### PR TITLE
pythonPackages.osqp: remove mkl dependency

### PR DESCRIPTION
**Pull request by drewrisinger**
###### Motivation for this change
Removing mkl allows this to be built in Hydra (mkl is unfree-ish license), and it seems the mkl dependency is somewhat optional given downstream packages build cleanly when removing mkl.
Also add explicit numpy/scipy dependency.
Cleanup commit slightly for formatting.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by drewrisinger** [comment]
@GrahamcOfBorg build python37Packages.osqp python37Packages.cvxpy python38Packages.osqp python38Packages.cvxpy

File: pkgs/development/python-modules/osqp/default.nix [hunk 1]
```diff
@@ -23,16 +22,15 @@ buildPythonPackage rec {
   dontUseCmakeConfigure = true;
 
   propagatedBuildInputs = [
-    numpy
     future
+    numpy
```
Comment (FRidh) [inline pkgs/development/python-modules/osqp/default.nix:26]: Why add these new dependencies? Are they required for the package?

**Comment by bhipple** [comment]
CC @matthewbauer, who is looking at making this type of thing parameterizable in https://github.com/NixOS/nixpkgs/pull/83888

**Comment by drewrisinger** [comment]
They seem to be used in the code, it's just not listed in requirements
which I used when originally packaging.

**Comment by matthewbauer** [comment]
This case is a little different because osqp doesn't use MKL for BLAS, but MKL PARDISO (https://software.intel.com/en-us/mkl-developer-reference-fortran-intel-mkl-pardiso-parallel-direct-sparse-solver-interface) which is unique to MKL.

Anyway, it looks like this library just needs MKL to test the MKL functionality. Since osqp just uses dlopen, you can always enable MKL support through `LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$(nix-build '<nixpkgs>' -A mkl)/lib`.

**Review by matthewbauer — APPROVED** [review]

**Comment by drewrisinger** [comment]
> This case is a little different because osqp doesn't use MKL for BLAS, but MKL PARDISO (https://software.intel.com/en-us/mkl-developer-reference-fortran-intel-mkl-pardiso-parallel-direct-sparse-solver-interface) which is unique to MKL.
> 
> Anyway, it looks like this library just needs MKL to test the MKL functionality. Since osqp just uses dlopen, you can always enable MKL support through `LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$(nix-build '<nixpkgs>' -A mkl)/lib`.

Ack, thanks for the good tips. I wasn't familiar with what MKL PARDISO did. ``pythonPackages.osqp`` does just use MKL PARDISO for testing PARDISO functionality AFAICT, it probably also enables some runtime functionality but I view that as a downstream thing. I prefer for build purposes to have no unfree dependencies so as much as possible can be built/cached by Hydra.

File: pkgs/development/python-modules/osqp/default.nix (hunk 1 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/osqp/default.nix:26]: Re numpy/scipy: They seem to be used in the code, it's just not listed in requirements.txt
which I used when originally packaging. It seems to still build/test fine, probably due to them being added to PYTHONPATH by some upstream package.

Comment (drewrisinger) [reply pkgs/development/python-modules/osqp/default.nix:26]: re scipy: it'd originally been a test dependency, but I wanted to move it to a regular dependency b/c I found the above code in it.

**Comment by drewrisinger** [comment]
@GrahamcOfBorg build python37Packages.qiskit-aer python37Packages.qiskit-ignis
@GrahamcOfBorg build python38Packages.qiskit-aer python38Packages.qiskit-ignis

**Comment by drewrisinger** [comment]
Result of ``nixpkgs-review pr 84124``

9 package built:
python27Packages.osqp python37Packages.cvxpy python37Packages.osqp python37Packages.qiskit-aer python37Packages.qiskit-ignis python38Packages.cvxpy python38Packages.osqp python38Packages.qiskit-aer python38Packages.qiskit-ignis

**Review by bhipple — APPROVED** [review]
Result of nixpkgs-review pr 84124 [1](https://github.com/Mic92/nixpkgs-review)<details>  <summary>9 package built:</summary>  - python27Packages.osqp  - python37Packages.cvxpy  - python37Packages.osqp  - python37Packages.qiskit-aer  - python37Packages.qiskit-ignis  - python38Packages.cvxpy  - python38Packages.osqp  - python38Packages.qiskit-aer  - python38Packages.qiskit-ignis</details>

**Comment by bhipple** [comment]
Can you send a 20.03 backport PR as well?

**Comment by drewrisinger** [comment]
@bhipple python3Packages.osqp isn't in 20.03.

**Comment by bhipple** [comment]
Ah ok, my bad!  For some reason I thought we'd had this packaged for longer, but I was thinking of another related pkg.

**Pull request merged**